lib/AST/IRToASTVisitor.cpp

```diff
@@ -24,392 +27,885 @@
-                               IRToTypeDeclMap &type_decls,
-                               IRToValDeclMap &value_decls, IRToStmtMap &stmts,
-                               ArgToTempMap &temp_decls)
-    : provenance(provenance),
-      ast_ctx(unit.getASTContext()),
-      ast(unit),
-      type_decls(type_decls),
-      value_decls(value_decls),
-      temp_decls(temp_decls) {}
-
-clang::QualType IRToASTVisitor::GetQualType(llvm::Type *type) {
-  DLOG(INFO) << "GetQualType: " << LLVMThingToString(type);
-
-  clang::QualType result;
-  switch (type->getTypeID()) {
-    case llvm::Type::VoidTyID:
-      result = ast_ctx.VoidTy;
-      break;
-
-    case llvm::Type::HalfTyID:
-      result = ast_ctx.HalfTy;
-      break;
-
-    case llvm::Type::FloatTyID:
-      result = ast_ctx.FloatTy;
-      break;
-
-    case llvm::Type::DoubleTyID:
-      result = ast_ctx.DoubleTy;
-      break;
-
-    case llvm::Type::X86_FP80TyID:
-      result = ast_ctx.LongDoubleTy;
-      break;
-
-    case llvm::Type::IntegerTyID: {
-      auto size{type->getIntegerBitWidth()};
-      CHECK(size > 0) << "Integer bit width has to be greater than 0";
-      result = ast.GetLeastIntTypeForBitWidth(size, /*sign=*/0);
-    } break;
-
-    case llvm::Type::FunctionTyID: {
-      auto func{llvm::cast<llvm::FunctionType>(type)};
-      auto ret{GetQualType(func->getReturnType())};
-      std::vector<clang::QualType> params;
-      for (auto param : func->params()) {
-        params.push_back(GetQualType(param));
-      }
-      auto epi{clang::FunctionProtoType::ExtProtoInfo()};
-      epi.Variadic = func->isVarArg();
-      result = ast_ctx.getFunctionType(ret, params, epi);
-    } break;
-
-    case llvm::Type::PointerTyID: {
-      auto ptr{llvm::cast<llvm::PointerType>(type)};
-      result = ast_ctx.getPointerType(GetQualType(ptr->getElementType()));
-    } break;
-
-    case llvm::Type::ArrayTyID: {
-      auto arr{llvm::cast<llvm::ArrayType>(type)};
-      auto elm{GetQualType(arr->getElementType())};
-      result = GetConstantArrayType(ast_ctx, elm, arr->getNumElements());
-    } break;
-
-    case llvm::Type::StructTyID: {
-      clang::RecordDecl *sdecl{nullptr};
-      auto &decl{type_decls[type]};
-      if (!decl) {
-        auto tudecl{ast_ctx.getTranslationUnitDecl()};
-        auto strct{llvm::cast<llvm::StructType>(type)};
-        auto sname{strct->isLiteral() ? ("literal_struct_" +
-                                         std::to_string(num_literal_structs++))
-                                      : strct->getName().str()};
-        if (sname.empty()) {
-          sname = "struct" + std::to_string(num_declared_structs++);
+  ASTBuilder ast;
+
+  Provenance &provenance;
+  size_t num_literal_structs = 0;
+  size_t num_declared_structs = 0;
+
+ public:
+  ExprGen(clang::ASTUnit &unit, Provenance &provenance)
+      : ast_ctx(unit.getASTContext()), ast(unit), provenance(provenance) {}
+
+  clang::QualType GetQualType(llvm::Type *type) {
```

Comment:
Can we factor out method bodies like `GetQualType` out of the class declaration? Again, no other but for consistency and/or neatness's sake. I promise this is the last piece of annoyance :eyes:

Reply:
Do you mean making them `static` or separating all the declarations of `ExprGen` and `StmtGen` from their definitions?